api/client/src/testFixtures/java/org/projectnessie/client/auth/oauth2/ResourceOwnerEmulator.java

```diff
@@ -0,0 +1,227 @@
+      authUrlConn.setRequestMethod("GET");
+      HttpURLConnection conn = expectLoginPage ? doLogin(authUrlConn) : authUrlConn;
+      conn.setInstanceFollowRedirects(false);
+      int status = conn.getResponseCode();
+      assertThat(status).isIn(HTTP_MOVED_TEMP, HTTP_MOVED_PERM);
+      invokeCallbackUrl(conn.getHeaderField("Location"));
+    } catch (Exception | AssertionError t) {
+      recordFailure(t);
+    }
+  }
+
+  /** Emulate user logging in to the authorization server. Note: this is Keycloak-specific. */
+  private HttpURLConnection doLogin(HttpURLConnection authUrlConn) throws IOException {
+    String html;
+    try (InputStream is = authUrlConn.getInputStream()) {
+      html =
+          new BufferedReader(new InputStreamReader(is, UTF_8))
+              .lines()
+              .collect(Collectors.joining("\n"));
+    }
+    Matcher matcher = KC_AUTH_FORM_PATTERN.matcher(html);
+    assertThat(matcher.find()).isTrue();
+    URL loginUrl = new URL(matcher.group(1));
+    // replace hostname and port with those from the auth URL;
+    // this is necessary because the auth server may be sending
+    // the login form to its configured frontend hostname + port,
+    // which, usually when using Docker, is something like keycloak:8080
+    loginUrl =
+        new URL(
+            loginUrl.getProtocol(),
+            authUrlConn.getURL().getHost(),
+            authUrlConn.getURL().getPort(),
+            loginUrl.getFile());
+    List<String> cookies = authUrlConn.getHeaderFields().get("Set-Cookie");
+    // send login form
+    HttpURLConnection loginUrlConn = (HttpURLConnection) loginUrl.openConnection();
+    loginUrlConn.setRequestMethod("POST");
+    for (String c : cookies) {
+      loginUrlConn.addRequestProperty("Cookie", c);
+    }
+    loginUrlConn.setDoOutput(true);
+    loginUrlConn.getOutputStream().write(loginData);
+    loginUrlConn.getOutputStream().close();
+    return loginUrlConn;
+  }
+
+  /** Emulate browser being redirected to callback URL. */
+  private void invokeCallbackUrl(String location) throws IOException {
+    URL callbackUrl = new URL(location);
+    assertThat(callbackUrl)
+        .hasPath(AuthorizationCodeFlow.CONTEXT_PATH)
+        .hasParameter("code")
+        .hasParameter("state");
+    HttpURLConnection con = (HttpURLConnection) callbackUrl.openConnection();
+    con.setRequestMethod("GET");
+    int status = con.getResponseCode();
+    assertThat(status).isEqualTo(HTTP_OK);
+  }
+
+  private void recordFailure(Throwable t) {
+    if (!closing) {
+      Consumer<Throwable> errorListener = this.errorListener;
+      if (errorListener != null) {
+        errorListener.accept(t);
+      }
+      Throwable e = error;
+      if (e == null) {
+        error = t;
+      } else {
+        e.addSuppressed(t);
+      }
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    closing = true;
+    System.setOut(standardOut);
```

Comment:
we should probably do this only when `replaceSystemOut == true`... In theory somebody else may be overriding `System.out` too.